sdk/python/feast/permissions/security_manager.py

```diff
@@ -102,8 +103,12 @@ def assert_permissions(
     Raises:
         PermissionError: If the current user is not authorized to execute the requested actions on the given resources.
     """
+    intra_communication_base64 = os.getenv("INTRA_COMMUNICATION_BASE64")
     sm = get_security_manager()
-    if sm is None:
+    if sm is None or (
```

Comment:
If the goal is to reduce the performance penalty in intra-server communication as much as possible, I think we will have to go a bit further and expose a method like `is_auth_necessary` that servers will call directly. In our design, servers sometimes need to fetch resources from the registry simply to pass them to security manager and they will still incur that penalty even with this additional if check.

Reply:
The goal in general is not to check permissions in intra server communication.
I tried to look for the scenario you were saying but I couldn't find any, can you please give me an example in the code?

Reply:
> If the goal is to reduce the performance penalty in intra-server communication as much as possible, I think we will have to go a bit further and expose a method like `is_auth_necessary` that servers will call directly. In our design, servers sometimes need to fetch resources from the registry simply to pass them to security manager and they will still incur that penalty even with this additional if check.

@tokoko The goal in general is not to check permissions in intra server communication.
I tried to look for the scenario you were saying but I couldn't find any, can you please give me an example in the code?

Reply:
Looks like there aren't many currently, but that's mostly because we have a bug for apply method permission checking (#4444).

btw, I'm fine with proceeding with the current approach for now and optimizing it later as well. Just pointing it out that this way there still might be a performance impact.

Reply:
@tokoko agree with prepending the asserts with this `is_auth_necessary` flag, now that [you explained](https://github.com/feast-dev/feast/issues/4444#issuecomment-2309822627) the reasons for the double asserts in the `apply` methods

Reply:
@tmihalac can we introduce the `is_auth_necessary` function and replace this `if` blocks?

Reply:
I think @dmartinol also meant to use `is_auth_necessary` in the servers to avoid unnecessary calculations, right?

Reply:
@tokoko I missed adding `is_auth_necessary` in newly added method `assert_permissions_to_update` in `security_manager.py`  is this what you meant?

Reply:
yes, exactly